### PR TITLE
Support nested results

### PR DESCRIPTION
**Pull request by blaix**
This builds in support for our usual style of paginated endpoints which return json like:

```
    {
        'results': [...],
        'next': ...
        ...
    }
```

ping @RayeN @jslang

### CODE REVIEW

**Comment by jslang** [comment]
This looks good, but only question I have is why add this instead of letting the user define through whatever callable they pass in how to interpret results? For example:

```python
def paginated_fn(page, page_size):
    return real_paginated_fn(page, page_size)['i_put_results']['wherever_i_want']

PaginatedAPIIterator(paginated_fn)
```

This provides more flexibility and requires no knowledge from `PaginatedAPIIterator` on how to interact with the results.

**Comment by blaix** [comment]
Actually the more I think about it the more I think this should be the
default. Our use case is api endpoints, where a dict is actually the
recommended way to return paginated endpoints. What do you think?

**Comment by RayeN** [comment]
>  This provides more flexibility and requires no knowledge from PaginatedAPIIterator on how to interact with the results.

Usually we already have a service method which returns pagination meta along with service result (to be able to deal with pagination manually or get resource count, etc.).  You proposal means that we should either prepare an additional callable which strips pagination meta or change current service methods (which means we loose some flexibility of service method when we want to use it apart from `PaginatedAPIIterator `).  Also, the functionality @blaix adds doesn't restrict anyone from passing callable which already knows where to get results.

> Actually the more I think about it the more I think this should be the
default.

+1 to make this a default behaviour.  Maybe also accept `None` as special value to treat top-level JSON as list of resources (if there will be any case for that).

File: demands/pagination.py [hunk 1]
```diff
@@ -61,6 +61,19 @@ class PaginatedAPIIterator(object):
         >>> list(iterator)
         [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ... 99]
 
+    If your paginated function returns a dict instead of a list of results,
+    you can specify a `results_key` option.
+
+
+        >>> def numbers(page, page_size):
```
Comment (RayeN) [inline demands/pagination.py]: `get_numbers` ?

Comment (blaix) [inline demands/pagination.py]: I'm following the pattern in the rest of the docstring

**Comment by blaix** [comment]
I made it the default: 370155d1e9c19aa9b042037149eb67bdb8a41440

**Comment by thomasw** [comment]
> where a dict is actually the
> recommended way to return paginated endpoints.

This is only tangentially related, but I'll say it again since I haven't in a while: In general, dicts should wrap all JSON API responses. If you don't do this, it becomes impossible to later add additional data to a given endpoint in a backwards compatible way. It's really unfortunate that, in older versions at least, DRF returns a bare list unless you enable pagination. People using DRF should be doing whatever they can to keep their endpoints from generating such responses.

File: demands/pagination.py [hunk 2]
```diff
@@ -67,6 +79,7 @@ class PaginatedAPIIterator(object):
         PAGE_SIZE_PARAM: 'page_size',
         PAGE_SIZE: 100,
         PAGINATION_TYPE: PaginationType.PAGE,
+        'results_key': 'results',
```
Comment (RayeN) [inline demands/pagination.py]: Why 'results_key' isn't a constant?

Comment (blaix) [inline demands/pagination.py]: fixed

**Comment by RayeN** [comment]
And you still want to introduce START_PAGE=1 param in the scope of this pull, right?

**Comment by blaix** [comment]
Yes, I have another pull ready for that change once this is merged.

File: demands/pagination.py [hunk 3]
```diff
@@ -90,7 +104,14 @@ def _get_page(self, page):
             self.options[PAGE_PARAM]: page,
             self.options[PAGE_SIZE_PARAM]: self.options[PAGE_SIZE],
         })
-        return self.paginated_fn(*self.args, **kwargs)
+        return self._get_results(**kwargs)
+
+    def _get_results(self, **kwargs):
+        results = self.paginated_fn(*self.args, **kwargs)
+        results_key = self.options.get('results_key')
```
Comment (RayeN) [inline demands/pagination.py]: Use constant here as well?

**Comment by RayeN** [comment]
:+1:

File: demands/pagination.py (hunk 3 above)
Comment (blaix) [inline demands/pagination.py]: oops, thanks. fixed here as well

**Comment by RayeN** [comment]
:+1:

File: demands/pagination.py [hunk 4]
```diff
@@ -15,32 +15,44 @@ class PaginationType(object):
 class PaginatedAPIIterator(object):
     """Paginated API iterator
 
-    Provides an interator for items in a paginated function, useful for service
```
Comment (stefanor) [inline demands/pagination.py]: aww, interators are the best

Comment (thomasw) [inline demands/pagination.py]: interator: an iterator that integrates each returned value into our hearts.

**Pull request merged**